storage/remote/codec.go

```diff
@@ -160,6 +160,123 @@ func FromQueryResult(sortSeries bool, res *prompb.QueryResult) storage.SeriesSet
 	}
 }
 
+// chunkedResponseSeriesSet implements storage.SeriesSet.
+type chunkedResponseSeriesSet struct {
+	stream        *ChunkedReader
+	currentChunks []prompb.Chunk
+	currentLabels []prompb.Label
+	httpResp      *http.Response
+}
+
+func (c *chunkedResponseSeriesSet) Next() bool {
+	c.currentLabels = nil
+	c.currentChunks = nil
+	res := &prompb.ChunkedReadResponse{}
+	if err := c.stream.NextProto(res); err != nil {
+		if c.httpResp != nil {
```

Comment:
I think we should handle io.EOF differently then casual error, no? (:

Reply:
I am sorry. I dont understand what you mean. Can you please elaborate Bartek?

Reply:
Sure, This special error sentinel means EOF - End of File

There is good explanation here: https://golang.org/pkg/io/#pkg-variables

Reply:
Why this could be nil?  What if error is non-nil and not io EOF? It would be nice to return it in SeriesSet.Err right? (:

Reply:
This was done in order to address your comments on: about closing the connection.
```
if httpResp.StatusCode/100 != 2 {
		io.Copy(ioutil.Discard, httpResp.Body)
		httpResp.Body.Close()
```

Reply:
`c.httpResp` is null in the test case (`storage/remote/codec_test.go`) where the object is initialized as `ss := &chunkedResponseSeriesSet{stream: r}`

Reply:
[This](https://github.com/prometheus/prometheus/blob/master/storage/remote/chunked.go#L108) about NextProro() states that the error type can either be `io.EOF` or `io.ErrUnexpectedEOF`  Relating to your comment on `SeriesSet.Err`, do you mean something like this?
```
err := c.stream.NextProto(res)
if err!= nil {
     c.nextChunkError = err
}
```
and then
```
func (c *chunkedResponseSeriesSet) Err() error {
	return c.nextChunkError
}
```

The above pattern breaks some test cases in `TestEndpoints -v ./web/api/v1/` I'm investigating it....

Reply:
Hey the main problem with current code is that when you have error from ` c.stream.NextProto(res)` it's never handled. That's a problem. It always return true.

I would definitely except something like you propose in https://github.com/prometheus/prometheus/pull/7354/files#r487503100 just `c.err` would be enough.

What you need though is: 

```
if err := c.stream.NextProto(res); err!= nil {
     if err != io.EOF{
         c.err = err
     }
     return false
}
```

Reply:
> This link about NextProto() states that the error type can either be io.EOF or io.ErrUnexpectedEOF 

This does not matter what errors the current implementation can return. In next commit someone can return some other error (:

Reply:
Done. Please check